File: lectures/Python-module/04_Boost_Python/_5_properties.py
from _5_properties import Greetings

g = Greetings('Jacopo')

# greetings is readable
print(g.greetings) # this expression calls the getter 
                   # Hello Jacopo
# and writable
g.greetings = 'Alberto' # this expression calls the setter
print(g.greetings) # this expression calls the getter 
                   # Hello Alberto 

# name2greet is readable and writable
print(g.name2greet) # Jacopo
# but not writable 
try:
    g.name2greet = 'Alberto' # no setter for this property
except AttributeError as err: 
    print('ERR:',err,'| missing setter for name2greet property')

File: lectures/Python-module/04_Boost_Python/_5_properties.cc
#include <iostream>
struct Greetings
{
	Greetings(std::string name): name2greet(name) {this->set_greetings(name);} 
    void set_greetings(std::string name) {this->greetings = "Hello "+name;}
	std::string get_greetings() {return this->greetings;} 
	std::string get_name() {return this->name2greet;}
    std::string name2greet;
    std::string greetings;
};

#include <boost/python.hpp>
using namespace boost::python;

BOOST_PYTHON_MODULE(_5_properties) {
	class_<Greetings>("Greetings",init<std::string>())
	.add_property("name2greet",&Greetings::get_name) // only getter 
	.add_property("greetings" ,&Greetings::get_greetings  
			                  ,&Greetings::set_greetings) // setter/getter 
	;
}
